pkg/operator/ceph/cluster/watcher.go

```diff
@@ -130,6 +144,243 @@ func (c *clientCluster) onK8sNode(ctx context.Context, object runtime.Object) bo
 	return false
 }
 
+func (c *clientCluster) handleNodeFailure(ctx context.Context, cluster *cephv1.CephCluster, node *v1.Node) error {
+
+	watchForNodeLoss, err := k8sutil.GetOperatorSetting(ctx, c.context.Clientset, opcontroller.OperatorSettingConfigMapName, "ROOK_WATCH_FOR_NODE_FAILURE", "true")
+	if err != nil {
+		return fmt.Errorf("failed to get configmap value `ROOK_WATCH_FOR_NODE_FAILURE`. %v", err)
+	}
+	if watchForNodeLoss != "true" {
+		logger.Debugf("not watching for node failures `ROOK_WATCH_FOR_NODE_FAILURE` is set to %q", watchForNodeLoss)
+		return nil
+	}
+
+	nodeHasOutOfServiceTaint := false
+	for _, taint := range node.Spec.Taints {
+		if taint.Key == v1.TaintNodeOutOfService {
+			nodeHasOutOfServiceTaint = true
+			logger.Debugf("Found taint: Key=%v, Value=%v\n", taint.Key, taint.Value)
+			break
+		}
+
+		logger.Debugf("node taint keys: %v", node.Spec.Taints)
+		logger.Debugf("node addresses: %v\n", node.Status.Addresses)
+	}
+
+	if !nodeHasOutOfServiceTaint {
+		logger.Infof("node %s does not have taint %s, skip handling node failure", node.Name, v1.TaintNodeOutOfService)
+		err := c.unfenceAndDeleteNetworkFence(ctx, *node, cluster)
+		if err != nil {
+			return fmt.Errorf("failed to delete network fence for node %q. %v", node.Name, err)
+		}
+		return nil
+	}
+	_, err = c.context.ApiExtensionsClient.ApiextensionsV1().CustomResourceDefinitions().Get(ctx, "networkfences.csiaddons.openshift.io", metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("networkfences.csiaddons.openshift.io CRD not found, skip creating networkFence")
+			return nil
+		}
+		return fmt.Errorf("failed to get CRD networkfences.csiaddons.openshift.io, skip creating networkFence. %v", err)
+	}
+
+	if nodeHasOutOfServiceTaint {
+		err := c.getIpAndcreateNetworkFence(ctx, *node, cluster)
+		if err != nil {
+			return fmt.Errorf("failed to create network fence for node %q. %v", node.Name, err)
+		}
+	}
+
+	return nil
+}
+
+func (c *clientCluster) getIpAndcreateNetworkFence(ctx context.Context, node corev1.Node, cluster *cephv1.CephCluster) error {
```

Comment:
simple names are sometimes the best
```suggestion
func (c *clientCluster) fenceNode(ctx context.Context, node corev1.Node, cluster *cephv1.CephCluster) error {
```

Reply:
actually, I'm having for short of idea for name the func and logging messages 😅